src/include/gurt/hash.h

```diff
@@ -539,7 +540,7 @@ struct d_hlink_ops {
 
 struct d_rlink {
 	d_list_t		rl_link;
-	uint32_t		rl_ref;
+	ATOMIC uint32_t		rl_ref;
```

Comment:
@jolivier23 any idea how we could fix build failure here? we need HAVE_STDATOMIC macro check work in the header file.

Reply:
Seems a g++ problem, following build failure on high version while building src/tests/ftest/cart/utest/test_linkage.cpp
`/usr/lib/gcc/x86_64-redhat-linux/8/include/stdatomic.h:230:49: error: ‘atomic_flag’ does not name a type; did you mean ‘atomic_load’?
 extern _Bool atomic_flag_test_and_set (volatile atomic_flag *);
                                                 ^~~~~~~~~~~
                                                 atomic_load
/usr/lib/gcc/x86_64-redhat-linux/8/include/stdatomic.h:233:58: error: ‘atomic_flag’ does not name a type; did you mean ‘atomic_load’?
 extern _Bool atomic_flag_test_and_set_explicit (volatile atomic_flag *,
                                                          ^~~~~~~~~~~
                                                          atomic_load
/usr/lib/gcc/x86_64-redhat-linux/8/include/stdatomic.h:238:41: error: ‘atomic_flag’ does not name a type; did you mean ‘atomic_load’?
 extern void atomic_flag_clear (volatile atomic_flag *);
                                         ^~~~~~~~~~~
                                         atomic_load
/usr/lib/gcc/x86_64-redhat-linux/8/include/stdatomic.h:240:50: error: ‘atomic_flag’ does not name a type; did you mean ‘atomic_load’?
 extern void atomic_flag_clear_explicit (volatile atomic_flag *, memory_order);
                                                  ^~~~~~~~~~~
                                                  atomic_load
In file included from src/include/gurt/hash.h:22,
                 from src/tests/ftest/cart/utest/test_linkage.cpp:15:
src/include/gurt/atomic.h:19:16: error: ‘_Atomic’ does not name a type
 #define ATOMIC _Atomic
`

Reply:
Currently problem is C++ did not support _ATOMIC keyword, so test_linkage.cpp break because it need include gurt/hash.h which i added a ATOMIC there, i think even without this PR, c++ break too, just try to include gurt/atomic.h in your C++ file will break build unfortunately.

I don't have idea how we could fix this(maybe rename .cpp to .c) for now to walkaround issue?

Reply:
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=60932

_ATOMIC only supported in c++20

Reply:
For now i walkaround issue to rename cpp to c to process testing.

Reply:
Maybe we need the extern(C) code in atomic.h?

Reply:
I tried that, it did not work unfortunately.

Reply:
Yup, this way looks working, thanks!